pkg/loghttp/push/otlp.go

```diff
@@ -0,0 +1,373 @@
+			)
+	}
+
+	return req.Logs(), nil
+}
+
+func otlpToLokiPushRequest(ld plog.Logs, userID string, tenantsRetention TenantsRetention, stats *Stats) *logproto.PushRequest {
+	if ld.LogRecordCount() == 0 {
+		return &logproto.PushRequest{}
+	}
+
+	rls := ld.ResourceLogs()
+	pushRequestsByStream := make(map[string]logproto.Stream, rls.Len())
+
+	for i := 0; i < rls.Len(); i++ {
+		sls := rls.At(i).ScopeLogs()
+		res := rls.At(i).Resource()
+
+		flattenedResourceAttributes := labels.NewBuilder(logproto.FromLabelAdaptersToLabels(attributesToLabels(res.Attributes(), "")))
+		// service.name is a required Resource Attribute. If it is not present, we will set it to "unknown_service".
+		if flattenedResourceAttributes.Get("service_name") == "" {
+			flattenedResourceAttributes = flattenedResourceAttributes.Set("service_name", "unknown_service")
+		}
+
+		if dac := res.DroppedAttributesCount(); dac != 0 {
+			flattenedResourceAttributes = flattenedResourceAttributes.Set("resource_dropped_attributes_count", fmt.Sprintf("%d", dac))
+		}
+
+		// copy blessed attributes to stream labels
+		streamLabels := make(model.LabelSet, len(blessedAttributesNormalized))
+		for _, ba := range blessedAttributesNormalized {
+			v := flattenedResourceAttributes.Get(ba)
+			if v == "" {
+				continue
+			}
+			streamLabels[model.LabelName(ba)] = model.LabelValue(v)
+
+			// remove the blessed attributes copied to stream labels
+			flattenedResourceAttributes.Del(ba)
+		}
+
+		if err := streamLabels.Validate(); err != nil {
+			stats.errs = append(stats.errs, fmt.Errorf("invalid labels: %w", err))
+			continue
+		}
+		labelsStr := streamLabels.String()
+
+		// convert the remaining resource attributes to structured metadata
+		resourceAttributesAsStructuredMetadata := logproto.FromLabelsToLabelAdapters(flattenedResourceAttributes.Labels())
+
+		lbs := modelLabelsSetToLabelsList(streamLabels)
+		if _, ok := pushRequestsByStream[labelsStr]; !ok {
+			pushRequestsByStream[labelsStr] = logproto.Stream{
+				Labels: labelsStr,
+			}
+			stats.streamLabelsSize += int64(labelsSize(logproto.FromLabelsToLabelAdapters(lbs)))
+		}
+
+		resourceAttributesAsStructuredMetadataSize := labelsSize(resourceAttributesAsStructuredMetadata)
+		stats.structuredMetadataBytes[tenantsRetention.RetentionPeriodFor(userID, lbs)] += int64(resourceAttributesAsStructuredMetadataSize)
+
+		for j := 0; j < sls.Len(); j++ {
+			scope := sls.At(j).Scope()
+			logs := sls.At(j).LogRecords()
+
+			// it would be rare to have multiple scopes so if the entries slice is empty, pre-allocate it for the number of log entries
+			if cap(pushRequestsByStream[labelsStr].Entries) == 0 {
+				stream := pushRequestsByStream[labelsStr]
+				stream.Entries = make([]push.Entry, 0, logs.Len())
+				pushRequestsByStream[labelsStr] = stream
+			}
+
+			// use fields and attributes from scope as structured metadata
+			scopeAttributesAsStructuredMetadata := attributesToLabels(scope.Attributes(), "")
```

Comment:
there is no way for the caller to know how big the slice has to be since the flattening of attributes could result in more elements when there are nested attributes. Also, passing that extra argument complicates the code. I would love to keep it simple and optimize it later if it happens to cause any with real usage.